run_action.py

```diff
@@ -495,6 +494,138 @@ def dismiss_change_requests(
         # Avoid triggering abuse detection
         time.sleep(10)
 
+    if auto_resolve_conversations:
+        resolve_conversations(
+            github_token=github_token,
+            repo=repo,
+            pull_request_id=pull_request_id,
+            github_api_timeout=github_api_timeout,
+        )
+
+
+def conversation_threads_to_close(repo, pr_number, github_token, github_api_timeout):
+    """Generator of unresolved conversation threads to close
+
+    Uses the GitHub GraphQL API to get conversation threads for the given PR.
+    Then filters for unresolved threads and those that have been created by the action.
+    """
+
+    repo_owner, repo_name = repo.split("/")
+    query = """
+    query {
+      repository(owner: "%s", name: "%s") {
+        pullRequest(number: %d) {
+          id
+          reviewThreads(last: 100) {
+            nodes {
+              id
+              isResolved
+              comments(first: 1) {
+                nodes {
+                  id
+                  body
+                  author {
+                    login
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    """ % (
+        repo_owner,
+        repo_name,
+        pr_number,
+    )
+
+    response = requests.post(
+        "https://api.github.com/graphql",
+        json={"query": query},
+        headers={"Authorization": "Bearer " + github_token},
+        timeout=github_api_timeout,
+    )
+
+    if response.status_code == 200:
+        data = response.json()
+
+        # Iterate through review threads
+        for thread in data["data"]["repository"]["pullRequest"]["reviewThreads"][
+            "nodes"
+        ]:
+            for comment in thread["comments"]["nodes"]:
+                if (
+                    comment["id"]
+                    and thread["isResolved"] is False
+                    # this actor here is somehow different from `github-actions[bot]`
+                    # which we get through the Rest API
+                    and comment["author"]["login"] == "github-actions"
+                ):
+                    yield thread
+                    break
+    else:
+        print(
+            f"::error::getting unresolved conversation threads: {response.status_code}"
+        )
+        raise RuntimeError("Failed to get unresolved conversation threads.")
```

Comment:
If the permission is missing, will we get a `403` or whatever here?
If yes, maybe add it to the printout as a hint to the user? :thinking:

Reply:
You'd expect that, yeah. But no. The response is still 200 in that case and the missing auth is only mentioned in the error string.